src/trailer.c

```diff
@@ -0,0 +1,415 @@
+	int in_old_conflicts_block = 0;
+	size_t cutoff = len;
+
+	while (bol < cutoff) {
+		const char *next_line = memchr(buf + bol, '\n', len - bol);
+
+		if (!next_line)
+			next_line = buf + len;
+		else
+			next_line++;
+
+		if (buf[bol] == comment_line_char || buf[bol] == '\n') {
+			/* is this the first of the run of comments? */
+			if (!boc)
+				boc = bol;
+			/* otherwise, it is just continuing */
+		} else if (starts_with(buf + bol, "Conflicts:\n")) {
+			in_old_conflicts_block = 1;
+			if (!boc)
+				boc = bol;
+		} else if (in_old_conflicts_block && buf[bol] == '\t') {
+			; /* a pathname in the conflicts block */
+		} else if (boc) {
+			/* the previous was not trailing comment */
+			boc = 0;
+			in_old_conflicts_block = 0;
+		}
+		bol = next_line - buf;
+	}
+	return boc ? len - boc : len - cutoff;
+}
+
+/*
+ * Return the position of the start of the patch or the length of str if there
+ * is no patch in the message.
+ */
+static int find_patch_start(const char *str)
+{
+	const char *s;
+
+	for (s = str; *s; s = next_line(s)) {
+		if (starts_with(s, "---"))
+			return s - str;
+	}
+
+	return s - str;
+}
+
+/*
+ * Return the position of the first trailer line or len if there are no
+ * trailers.
+ */
+static int find_trailer_start(const char *buf, size_t len)
+{
+	const char *s;
+	int end_of_title, l, only_spaces = 1;
+	int recognized_prefix = 0, trailer_lines = 0, non_trailer_lines = 0;
+	/*
+	 * Number of possible continuation lines encountered. This will be
+	 * reset to 0 if we encounter a trailer (since those lines are to be
+	 * considered continuations of that trailer), and added to
+	 * non_trailer_lines if we encounter a non-trailer (since those lines
+	 * are to be considered non-trailers).
+	 */
+	int possible_continuation_lines = 0;
+
+	/* The first paragraph is the title and cannot be trailers */
+	for (s = buf; s < buf + len; s = next_line(s)) {
+		if (s[0] == comment_line_char)
+			continue;
+		if (is_blank_line(s))
+			break;
+	}
+	end_of_title = s - buf;
+
+	/*
+	 * Get the start of the trailers by looking starting from the end for a
+	 * blank line before a set of non-blank lines that (i) are all
+	 * trailers, or (ii) contains at least one Git-generated trailer and
+	 * consists of at least 25% trailers.
+	 */
+	for (l = last_line(buf, len);
+	     l >= end_of_title;
+	     l = last_line(buf, l)) {
+		const char *bol = buf + l;
+		const char *const *p;
+		int separator_pos;
+
+		if (bol[0] == comment_line_char) {
+			non_trailer_lines += possible_continuation_lines;
+			possible_continuation_lines = 0;
+			continue;
+		}
+		if (is_blank_line(bol)) {
+			if (only_spaces)
+				continue;
+			non_trailer_lines += possible_continuation_lines;
+			if (recognized_prefix &&
+			    trailer_lines * 3 >= non_trailer_lines)
+				return next_line(bol) - buf;
+			else if (trailer_lines && !non_trailer_lines)
+				return next_line(bol) - buf;
+			return len;
+		}
+		only_spaces = 0;
+
+		for (p = git_generated_prefixes; *p; p++) {
```

Comment:
Why do we treat git generated prefixes differently? Trailer lines can be whatever information the user wants them to be, so in my opinion there is no need to check for git prefixes but instead to check for the expected trailer line format.

Reply:
My [comment below](https://github.com/libgit2/libgit2/pull/4451#discussion_r159573210) has some context around this but so you don't have to skip around (that comment is shown as outdated now for some reason) - https://github.com/git/git/commit/146245063e286d5a38b146bd5a38da958bd3957b has more context on this. tldr; a trailer block is determined using a heuristic of "regular" trailers and git-created trailers.

Reply:
This logic is taken directly from Git. It seems pretty funky to me, but the goal here is to be compatible with Git so I've brought over this git generated prefix logic as-is.

Reply:
Yeah, I definitly agree that we should strive to follow upstream-logic as close as possible. I really don't like all these workarounds and heuristics being put into so many places just because the initial design wasn't flexible enough. Nothing we can do about that... thanks though for the pointers!

Reply:
We may have not been very clear in the OP or throughout this discussion, but this isn't "workarounds and heuristics" of the initial design - it's [literally the upstream code itself](https://github.com/git/git/blob/master/trailer.c#L861). Which by my understanding is the source of truth for how any of this is supposed to work anyway, right?

The design of the code may not match the standards of libgit2, but if we (libgit2) are to handle trailers exactly the same way git does, we unfortunately need to also take the heuristics in to account whenever we refactor this code anyway.

For now at least, literally using the upstream code seems like the safest (at least in terms of compatibility) way to go.

Reply:
Sorry for not making myself clear, but with "initial design" I referred to upstream git.git, not your code. So I do not complain about your implementation but all the heuristics employed by git.git.